File: Upnp/Public/Cs/TestPerformanceCp.cs
using System;
using System.Threading;
using System.Collections.Generic;
using OpenHome.Net.Core;
using OpenHome.Net.ControlPoint;
using OpenHome.Net.ControlPoint.Proxies;

namespace OpenHome.Net.Device
{
    class Program
    {
        class ThreadArgs
        {
            public CpDevice Device { get; private set; }
            public int Count
            {
                get { lock (this) { return iCount; } }
            }

            private DateTime iStopTime;
            private int iCount;

            public ThreadArgs(CpDevice aDevice, double aDurationMs)
            {
                Device = aDevice;
                iStopTime = DateTime.Now.AddMilliseconds(aDurationMs);
            }
            public bool IncrementCount()
            {
                lock (this)
                {
                    if (DateTime.Now <= iStopTime)
                    {
                        iCount++;
                        return true;
                    }
                    return false;
                }
            }
        }

        private static readonly int kTestDurationMs = 1000;

        public static void Main(string[] aArgs)
        {
            Console.WriteLine("TestPerformanceCpCs - starting");
            InitParams initParams = new InitParams();
            using (Library lib = Library.Create(initParams))
            {
                IntPtr subnetList = lib.SubnetListCreate();
                IntPtr nif = lib.SubnetAt(subnetList, 0);
                uint subnet = lib.NetworkAdapterSubnet(nif);
                lib.SubnetListDestroy(subnetList);
                var deviceListFactory = new CpUpnpDeviceListFactory(lib.StartCp(subnet));
                CpDevice device = null;
                Semaphore sem = new Semaphore(0, 1);
                var deviceList = deviceListFactory.CreateListServiceType("openhome.org", "TestBasic", 1,
                        (aDeviceList, aDevice) =>
                        {
                            if (device != null)
                            {
                                throw new Exception("Found more than one device.  Giving up as test results will probably be invalid.");
                            }
                            device = aDevice;
                            device.AddRef();
                            sem.Release();
                        },
                        (aDeviceList, aDevice) =>
                        {
                            throw new Exception("ERROR: Device removed while test is running.");
                        });
                sem.WaitOne();

                // actions
                Console.WriteLine("");
                int[] threadCounts = { 1, 2, 4 };
                foreach (int threadCount in threadCounts)
                {
                    List<Thread> threads = new List<Thread>();
                    for (int i = 0; i < threadCount; i++)
                    {
                        threads.Add(new Thread(ActionThread));
                    }
                    ThreadArgs threadArgs = new ThreadArgs(device, kTestDurationMs);
                    for (int i = 0; i < threadCount; i++)
                    {
                        threads[i].Start(threadArgs);
                    }
                    for (int i = 0; i < threadCount; i++)
                    {
                        threads[i].Join();
                    }
                    Console.WriteLine("Invoked {0} actions in {1}ms using {2} threads", threadArgs.Count, kTestDurationMs, threadCount);
                }

                // subscriptions
                Thread thread = new Thread(SubscriptionThread);
                ThreadArgs args = new ThreadArgs(device, kTestDurationMs);
                thread.Start(args);
                thread.Join();
                Console.WriteLine("\nCompleted {0} subscriptions in {1}ms\n", args.Count, kTestDurationMs);

                device.RemoveRef();
                deviceList.Dispose();
                Console.WriteLine("\nTests complete.  Press 'q' to exit.");
                while (Console.ReadKey(true).KeyChar != 'q') ;
            }
        }
        private static void ActionThread(object aArgs)
        {
            ThreadArgs args = (ThreadArgs)aArgs;
            using (CpProxyOpenhomeOrgTestBasic1 proxy = new CpProxyOpenhomeOrgTestBasic1(args.Device))
            {
                uint val;
                do
                {
                    proxy.SyncGetUint(out val);
                } while (args.IncrementCount());
            }
        }
        private static void SubscriptionThread(object aArgs)
        {
            ThreadArgs args = (ThreadArgs)aArgs;
            Semaphore sem = new Semaphore(0, 1);
            do
            {
                CpProxyOpenhomeOrgTestBasic1 proxy = new CpProxyOpenhomeOrgTestBasic1(args.Device);
                proxy.SetPropertyChanged(() => sem.Release());
                proxy.Subscribe();
                sem.WaitOne();
                proxy.Dispose();
            } while (args.IncrementCount());
        }
    }
}

File: OpenHome/Net/Bindings/Java/org/openhome/net/controlpoint/tests/TestPerformanceCp.java
package org.openhome.net.controlpoint.tests;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Inet4Address;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.Semaphore;

import org.openhome.net.controlpoint.CpDevice;
import org.openhome.net.controlpoint.CpUpnpDeviceListFactory;
import org.openhome.net.controlpoint.ICpDeviceList;
import org.openhome.net.controlpoint.ICpDeviceListListener;
import org.openhome.net.controlpoint.ICpUpnpDeviceListFactory;
import org.openhome.net.controlpoint.IPropertyChangeListener;
import org.openhome.net.controlpoint.proxies.CpProxyOpenhomeOrgTestBasic1;
import org.openhome.net.core.InitParams;
import org.openhome.net.core.Library;
import org.openhome.net.core.NetworkAdapter;
import org.openhome.net.core.SubnetList;

public class TestPerformanceCp implements ICpDeviceListListener
{
	public static final int TEST_DURATION_MS = 1000;
	
	private class ThreadArgs
    {
        private CpDevice iDevice;
        private int iCount;
        private Date iStopTime;

        public ThreadArgs(CpDevice aDevice, long aDurationMs)
        {
            iDevice = aDevice;
            iStopTime = new Date((new Date()).getTime() + aDurationMs);
        }
        public CpDevice getDevice()
        {
        	return iDevice;
        }
        public int getCount()
        {
        	synchronized (this)
        	{
        		return iCount;
        	}
        }
        public boolean incrementCount()
        {
            synchronized (this)
            {
            	Date timeNow = new Date();
                if (timeNow.before(iStopTime) || timeNow.equals(iStopTime))
                {
                    iCount++;
                    return true;
                }
                return false;
            }
        }
    }
	
	private class ActionThread extends Thread
    {
		private ThreadArgs iArgs;
		private CpProxyOpenhomeOrgTestBasic1 iProxy;
		
		public ActionThread(ThreadArgs aArgs)
		{
			iArgs = aArgs;
			iProxy = new CpProxyOpenhomeOrgTestBasic1(aArgs.getDevice());
		}

		public void run()
		{
            long val;
            do
            {
            	val = iProxy.syncGetUint();
            } while (iArgs.incrementCount());
		}
    }
    private class SubscriptionThread extends Thread implements IPropertyChangeListener
    {
        ThreadArgs iArgs;
        Semaphore iSem;
        
        public SubscriptionThread(ThreadArgs aArgs)
        {
        	iArgs = aArgs;
        	iSem = new Semaphore(1);
        	iSem.acquireUninterruptibly();
        }

        public void run()
        {
	        do
	        {
	        	CpProxyOpenhomeOrgTestBasic1 proxy = new CpProxyOpenhomeOrgTestBasic1(iArgs.getDevice());
	        	proxy.setPropertyChanged(this);
	            proxy.subscribe();
	            iSem.acquireUninterruptibly();
	            proxy.dispose();
	        } while (iArgs.incrementCount());
        }

		public void notifyChange() {
			iSem.release();
		}
    }
	
	private CpDevice iDevice;
	private Semaphore iSem;
	
	private TestPerformanceCp()
	{
		System.out.println("TestPerformanceCpJava - starting");
		InitParams initParams = new InitParams();
        Library lib = Library.create(initParams);
        SubnetList subnetList = new SubnetList();
        NetworkAdapter nif = subnetList.getSubnet(0);
        Inet4Address subnet = nif.getSubnet();
        subnetList.destroy();
        ICpUpnpDeviceListFactory deviceListFactory = new CpUpnpDeviceListFactory(lib.startCp(subnet));
        iDevice = null;
        iSem = new Semaphore(1);
    	iSem.acquireUninterruptibly();
        ICpDeviceList deviceList = deviceListFactory.createListServiceType("openhome.org", "TestBasic", 1, this);
        iSem.acquireUninterruptibly();
        
        // actions
        System.out.println("");
        int[] threadCounts = { 1, 2, 4 };
        for (int threadCount : threadCounts)
        {
            List<Thread> threads = new ArrayList<Thread>();
            ThreadArgs threadArgs = new ThreadArgs(iDevice, TEST_DURATION_MS);
            for (int i = 0; i < threadCount; i++)
            {
                threads.add(new ActionThread(threadArgs));
            }
            for (int i = 0; i < threadCount; i++)
            {
                threads.get(i).start();
            }
            for (int i = 0; i < threadCount; i++)
            {
            	try
            	{
            		threads.get(i).join();
            	}
            	catch (InterruptedException ie)
            	{
            		ie.printStackTrace();
            	}
            }
            System.out.printf("Invoked %d actions in %dms using %d threads\n", threadArgs.getCount(), TEST_DURATION_MS, threadCount);
        }

        // subscriptions
        ThreadArgs args = new ThreadArgs(iDevice, TEST_DURATION_MS);
        Thread thread = new SubscriptionThread(args);
        thread.start();
        try {
        	thread.join();
        }
        catch (InterruptedException ie)
        {
        	ie.printStackTrace();
        }
        System.out.printf("\nCompleted %d subscriptions in %dms\n", args.getCount(), TEST_DURATION_MS);

        iDevice.removeRef();
        deviceList.destroy();
        System.out.println("\nTests complete.  Press 'q' to exit.");
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        try
        {
        	while(in.read() != 'q') ;
        }
        catch (IOException io)
        {
        	System.err.println("ERROR: Exiting abnormally");
        }
	}

	public void deviceAdded(CpDevice aDevice) {
		if (iDevice != null)
        {
            throw new RuntimeException("Found more than one device.  Giving up as test results will probably be invalid.");
        }
        iDevice = aDevice;
        iDevice.addRef();
        iSem.release();
	}

	public void deviceRemoved(CpDevice aDevice) {
		throw new RuntimeException("ERROR: Device removed while test is running.");
	}
	
	public static void main(String[] args)
	{
        TestPerformanceCp testPerf = new TestPerformanceCp();
	}
}
